Sources/Safetensors/Safetensors.swift

```diff
@@ -283,31 +91,22 @@ public enum Safetensors {
         _ data: [String: any SafetensorsEncodable],
         metadata: [String: String]? = nil
     ) throws -> Data {
-        var headerData = [String: HeaderElement]()
-        headerData.reserveCapacity(data.count + (metadata == nil ? 0 : 1))
-        var previousOffset = 0
-        var tensorData = [UInt8]()
+        var headerData = ParsedSafetensors.HeaderData(
+            minimumCapacity: data.count + (metadata == nil ? 0 : 1)
+        )
+    
+        let totalDataSize = try data.values.reduce(0) { try $1.byteCount + $0 }
+        var tensorData: Data = .init(capacity: totalDataSize)
+        
         for (key, tensor) in data {
-            let tensorByteCount = try tensor.scalarSize() * tensor.scalarCount
-            let tensorHeaderData = try TensorData(
-                dtype: tensor.dtype(),
-                shape: tensor.tensorShape,
-                dataOffsets: OffsetRange(
-                    start: previousOffset,
-                    end: tensorByteCount + previousOffset
-                )
-            )
-            previousOffset += tensorByteCount
-            headerData[key] = .tensorData(tensorHeaderData)
-            try tensorData.append(contentsOf: tensor.toData())
+            headerData[key] = .tensorData(try tensor.tensorData(at: tensorData.count))
+            tensorData.append(try tensor.toData())
         }
+        
         if let metadata {
             headerData["__metadata__"] = .metadata(metadata)
         }
-        let encoder = JSONEncoder()
-        encoder.keyEncodingStrategy = .convertToSnakeCase
-        let header = try encoder.encode(headerData)
-        let headerSize = withUnsafeBytes(of: UInt64(header.count)) { Data($0) }
-        return headerSize + header + Data(tensorData)
+
+        return try HeaderEncoder().encode(headerData) + tensorData
```

Comment:
I think there is no need to instantiate a struct just to encode some data, this can be just a free function I think